arraycontext/container/arithmetic.py

```diff
@@ -133,6 +133,7 @@ def with_container_arithmetic(
         matmul: bool = False,
         bitwise: bool = False,
         shift: bool = False,
+        _same_cls_check: Optional[str] = None,
```

Comment:
Not liking the idea of letting users inject arbitrary code in here (even more than with `*_init_arrays_code`, I guess?). Why not just always enforce the array context check?

Reply:
Yeah, it's yucky and ad-hoc, hence the underscoring. Maybe you can tell I wasn't super proud of the interface. :)

At the same time, it's not generically possible to find the array context from the array. (That's the price we pay for allowing un-wrapped array types like the pyopencl array or the numpy array.) Calling `get_container_context` would run into the same problem. I also would like to keep this code path fast. Those are the constraints from my point of view. I would absolutely welcome better ideas, but I'm not sure I have any.

Reply:
How about adding a `promise_this_class_has_an_array_context_attribute: bool = False` and generating the check based on that? (preferably with a better name)

Reply:
That is better. 21b1462

Reply:
Nice! :tada: